core/offchain-worker-executor/src/executor.rs

```diff
@@ -0,0 +1,274 @@
+	ParentchainBlock,
+	TopPoolAuthor,
+	StfExecutor,
+	StateHandler,
+	ValidatorAccessor,
+	ExtrinsicsFactory,
+> {
+	top_pool_author: Arc<TopPoolAuthor>,
+	stf_executor: Arc<StfExecutor>,
+	state_handler: Arc<StateHandler>,
+	validator_accessor: Arc<ValidatorAccessor>,
+	extrinsics_factory: Arc<ExtrinsicsFactory>,
+	_phantom: PhantomData<ParentchainBlock>,
+}
+
+impl<
+		ParentchainBlock,
+		TopPoolAuthor,
+		StfExecutor,
+		StateHandler,
+		ValidatorAccessor,
+		ExtrinsicsFactory,
+	>
+	Executor<
+		ParentchainBlock,
+		TopPoolAuthor,
+		StfExecutor,
+		StateHandler,
+		ValidatorAccessor,
+		ExtrinsicsFactory,
+	> where
+	ParentchainBlock: Block<Hash = H256>,
+	StfExecutor: StateUpdateProposer,
+	TopPoolAuthor: AuthorApi<H256, ParentchainBlock::Hash>,
+	StateHandler: QueryShardState + HandleState<StateT = StfExecutor::Externalities>,
+	ValidatorAccessor: ValidatorAccess<ParentchainBlock> + Send + Sync + 'static,
+	ExtrinsicsFactory: CreateExtrinsics,
+	NumberFor<ParentchainBlock>: BlockNumberOps,
+{
+	pub fn new(
+		top_pool_author: Arc<TopPoolAuthor>,
+		stf_executor: Arc<StfExecutor>,
+		state_handler: Arc<StateHandler>,
+		validator_accessor: Arc<ValidatorAccessor>,
+		extrinsics_factory: Arc<ExtrinsicsFactory>,
+	) -> Self {
+		Self {
+			top_pool_author,
+			stf_executor,
+			state_handler,
+			validator_accessor,
+			extrinsics_factory,
+			_phantom: Default::default(),
+		}
+	}
+
+	pub fn execute(&self) -> Result<()> {
+		let max_duration = Duration::from_secs(5);
+		let latest_parentchain_header = self.get_latest_parentchain_header()?;
+
+		let mut parentchain_effects: Vec<OpaqueCall> = Vec::new();
+
+		let shards = self.state_handler.list_shards()?;
+		debug!("Executing calls on {} shard(s)", shards.len());
+
+		for shard in shards {
+			let trusted_calls = self.top_pool_author.get_pending_tops_separated(shard)?.0;
+			debug!("Executing {} trusted calls on shard {:?}", trusted_calls.len(), shard);
+
+			let batch_execution_result = self.stf_executor.propose_state_update(
+				&trusted_calls,
+				&latest_parentchain_header,
+				&shard,
+				max_duration,
+				|s| s,
+			)?;
+
+			parentchain_effects
+				.append(&mut batch_execution_result.get_extrinsic_callbacks().clone());
+
+			let failed_operations = batch_execution_result.get_failed_operations();
+			let successful_operations: Vec<ExecutedOperation> = batch_execution_result
+				.get_executed_operation_hashes()
+				.into_iter()
+				.map(|h| ExecutedOperation::success(h, TrustedOperationOrHash::Hash(h), Vec::new()))
+				.collect();
+
+			// Remove all not successfully executed operations from the top pool.
+			self.remove_calls_from_pool(&shard, failed_operations);
+
+			// Apply the state update
+			self.apply_state_update(&shard, batch_execution_result.state_after_execution)?;
+
+			// Remove successful operations from pool
+			self.remove_calls_from_pool(&shard, successful_operations);
+
+			// TODO: notify parentchain about executed operations? -> add to parentchain effects
```

Comment:
I believe that fluency asked for this, and it makes sense so you can verify on-chain that your transaction was executed?

Reply:
Yes, this could be nice. However, if we do this, we need to do it twice; once for `inBlock` and once for `finalized`. However, we could also just add all executed calls into the roll-up information: https://github.com/integritee-network/worker/issues/855